Make Swift aware of blocks on Linux and other non-Darwin systems.

al45tair opened the pull request:
We need to add a new value witness table for blocks to make this work, not least because we may not have ObjC interop, but we may still have block support.

rdar://82092046

al45tair committed:
[Tests] Remove guards preventing @convention(block) test from running on Linux

This should result in an assertion failure on the Linux build, which needs
to be fixed; blocks *are* supposed to be supported on Linux, apparently.

rdar://82092046

al45tair commented:
@swift-ci Please smoke test Linux platform

al45tair commented:
@swift-ci Please test Linux platform

al45tair commented:
Hmmm, OK. So it works on "Linux platform". Let's try some others

@swift-ci Please test CentOS 7 platform

al45tair commented:
@swift-ci Please test Centos 8 platform

al45tair commented:
@swift-ci Please test Ubuntu 18.04 platform

al45tair commented:
@swift-ci Please test Ubuntu 20.04 platform

al45tair commented:
@swift-ci Please test Amazon Linux 2 platform

al45tair commented:
@swift-ci Please smoke test Linux platform

CodaFi commented on test/TypeRoundTrip/Inputs/testcases/structural_types.swift:
```diff
@@ -24,9 +24,7 @@ public func test() {
   roundTripType(Array<@convention(c) () -> ()>.self)
 
   // @convention(block) requires Objective-C support
-  #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```
You can use `#if _runtime(_ObjC)` to check for this instead.

al45tair replied:
Indeed. It's still the wrong thing, though. This should work regardless of ObjC interop.

al45tair committed:
[Runtime] Support blocks on systems without ObjC interop.

Added a new value witness table for blocks, rather than trying to reuse the
one for ObjC objects.  Use it.

Also link against libBlocksRuntime in unittests, when required.

rdar://82092046

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair committed:
[Stdlib] Add the blocks runtime to the linker options for -lswiftCore.

We need the blocks runtime on Windows and Linux.

rdar://82092046

al45tair commented:
@swift-ci Please test
@swift-ci Please smoke test
